lib/pure/decimal.nim

```diff
@@ -0,0 +1,1040 @@
+
+#
+# constants
+#
+
+# private constants
+const
+  significandSize: int = 34
+  bias: int16 = 6176
+  expLowerBound: int16 = 0 - bias
+  tempSignificandSize: int = 70
+  billion: uint32 = 1000 * 1000 * 1000 # a billion has 10 digits and fits into 32 bits
+  tenThousand: uint16 = 10 * 1000      # 10 thousand has 5 digits and fits into 16 bits
+  allZeroes: SignificandArray = [0.byte, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
+  transientOffset = tempSignificandSize - significandSize
+  noValue: int16 = -32768
+
+
+# masks (private)
+const
+  # all flags/exponents are in the first 32 bits ('a'), so they
+  # are represented by the uint32
+  # 
+  # the exponent fits into 14 bits, but the mask determines the prefix, so you
+  # really only have a range of 3 x 2^12 = 12288. So the bias of 6176 splits
+  # that in half, thus fitting the official range of -6143 to +6144.
+  signMask =                   0b1000_0000_0000_0000_0000_0000_0000_0000'u32
+  #
+  comboMask =                  0b0111_1111_1111_1111_1000_0000_0000_0000'u32
+  #
+  comboShortMask =             0b0110_0000_0000_0000_0000_0000_0000_0000'u32
+  comboShortSignalsMedium =    0b0110_0000_0000_0000_0000_0000_0000_0000'u32
+  comboShortExponentMask =     0b0111_1111_1111_1110_0000_0000_0000_0000'u32  # 14 bits
+  comboShortExponentShiftR =   16 + 1
+  #
+  # medium is used when the leading bit in the significand happens to be 1
+  comboMediumMask =            0b0111_1000_0000_0000_0000_0000_0000_0000'u32
+  comboMediumSignalsLong =     0b0111_1000_0000_0000_0000_0000_0000_0000'u32
+  comboMediumExponentMask =    0b0001_1111_1111_1111_1000_0000_0000_0000'u32
+  comboMediumExponentShiftR =  15
+  #
+  comboLongMask =              0b0111_1100_0000_0000_0000_0000_0000_0000'u32
+  comboLongInfinityFlag =      0b0111_1000_0000_0000_0000_0000_0000_0000'u32
+  comboLongNanFlag =           0b0111_1100_0000_0000_0000_0000_0000_0000'u32
+  comboLongSignalingNanFlag =  0b0000_0010_0000_0000_0000_0000_0000_0000'u32
+  #
+  # NOTE: the spec allows for sig combo bits and we are following BID variant.
+  #
+  # the largest significand is:
+  #     9999999999999999999999999999999999 (decimal) or
+  #    0001ed09_bead87c0_378d8e63_ffffffff (hex)
+  # which fits neatly into 113 bits. So, the masks for the first 32 bits is:
+  significandMaskUpper      =  0b0000_0000_0000_0001_1111_1111_1111_1111'u32
+  significandMediumBit      =  0b0000_0000_0000_0001_0000_0000_0000_1111'u32
+  #
+  allOnes                   =  0b1111_1111_1111_1111_1111_1111_1111_1111'u32
+
+when not defined(js):
+  const
+    upperOnes               =  0xFFFFFFFF00000000'u64
+    lowerOnes               =  0x00000000FFFFFFFF'u64
+
+
+# public constants
+let
+  nan* = Decimal(a: comboLongNanFlag, b:0, c:0, d:0)
+  infinity* = Decimal(a: comboLongInfinityFlag, b: 0, c:0, d:0)
+  negativeInfinity* = Decimal(a: (comboLongInfinityFlag or signMask), b: 0, c:0, d:0)
```

Comment:
```suggestion
const
  nan* = Decimal(a: comboLongNanFlag, b:0, c:0, d:0)
  infinity* = Decimal(a: comboLongInfinityFlag, b: 0, c:0, d:0)
  negativeInfinity* = Decimal(a: (comboLongInfinityFlag or signMask), b: 0, c:0, d:0)
```
This works when you're defining `Decimal` without `of RootObj`.

Reply:
fixed in next commit